src/ast/analysis/Type.cpp

```diff
@@ -823,23 +840,282 @@ void TypeAnalysis::print(std::ostream& os) const {
+    }
+    fatal("Missing functor type.");
+}
+
+TypeAttribute TypeAnalysis::getFunctorArgType(const Functor* functor, const size_t idx) const {
+    assert(hasProcessedFunctor(functor) && "functor not yet processed");
+    if (auto* intrinsic = as<IntrinsicFunctor>(functor)) {
+        auto* info = functorInfo.at(intrinsic);
+        return info->params.at(info->variadic ? 0 : idx);
+    } else if (auto* udf = as<UserDefinedFunctor>(functor)) {
+        return udfDeclaration.at(udf->getName())->getArgsTypes().at(idx);
+    }
+    fatal("Missing functor type.");
+}
+
+const std::vector<TypeAttribute>& TypeAnalysis::getFunctorArgTypes(const UserDefinedFunctor& udf) const {
+    return udfDeclaration.at(udf.getName())->getArgsTypes();
+}
+
+bool TypeAnalysis::isStatefulFunctor(const UserDefinedFunctor* udf) const {
+    return udfDeclaration.at(udf->getName())->isStateful();
+}
+
+const std::map<const NumericConstant*, NumericConstant::Type>& TypeAnalysis::getNumericConstantTypes() const {
+    return numericConstantType;
+}
+
+bool TypeAnalysis::isMultiResultFunctor(const Functor& functor) {
+    if (isA<UserDefinedFunctor>(functor)) {
+        return false;
+    } else if (auto* intrinsic = as<IntrinsicFunctor>(functor)) {
+        auto candidates = functorBuiltIn(intrinsic->getBaseFunctionOp());
+        assert(!candidates.empty() && "at least one op should match");
+        return candidates[0].get().multipleResults;
+    }
+    fatal("Missing functor type.");
+}
+
+IntrinsicFunctors TypeAnalysis::validOverloads(const IntrinsicFunctor& func) const {
+    auto typeAttrs = [&](const Argument* arg) -> std::set<TypeAttribute> {
+        std::set<TypeAttribute> tyAttrs;
+        if (const auto* inf = dynamic_cast<const IntrinsicFunctor*>(arg)) {
+            if (hasProcessedFunctor(inf)) {
+                tyAttrs.insert(getFunctorReturnType(inf));
+                return tyAttrs;
+            }
+        }
+        auto&& types = getTypes(arg);
+        if (types.isAll())
+            return {TypeAttribute::Signed, TypeAttribute::Unsigned, TypeAttribute::Float,
+                    TypeAttribute::Symbol, TypeAttribute::Record};
+
+        for (auto&& ty : types)
+            tyAttrs.insert(getTypeAttribute(ty));
+        return tyAttrs;
+    };
+    auto retTys = typeAttrs(&func);
+    auto argTys = map(func.getArguments(), typeAttrs);
+
+    IntrinsicFunctors functorInfos = contains(functorInfo, &func)
+                                             ? functorBuiltIn(getPolymorphicOperator(&func))
+                                             : functorBuiltIn(func.getBaseFunctionOp());
+    auto candidates = filterNot(functorInfos, [&](const IntrinsicFunctorInfo& x) -> bool {
+        if (!x.variadic && argTys.size() != x.params.size()) return true;  // arity mismatch?
+        for (size_t i = 0; i < argTys.size(); ++i)
+            if (!contains(argTys[i], x.params[x.variadic ? 0 : i])) return true;
+        return !contains(retTys, x.result);
```

Comment:
This is getting hard to read, and does not match our usual standard
```suggestion
        // arity mismatch?
        if (!x.variadic && argTys.size() != x.params.size()) {
            return true;
        }
        for (size_t i = 0; i < argTys.size(); ++i) {
            if (!contains(argTys[i], x.params[x.variadic ? 0 : i])) {
                return true;
            }
        }
        return !contains(retTys, x.result);
```
It could be untangled further to make the logic in the nested ifs less convoluted